storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+    for (uint32_t xdes_n= 0; xdes_n < m_space->free_limit;
+         xdes_n+= FSP_EXTENT_SIZE)
+    {
+      /* Ignore doublewrite buffer extent */
+      if (m_space == fil_system.sys_space
+          && buf_dblwr.is_inside(xdes_n))
+        continue;
+      uint32_t descr_page_no=
+        xdes_calc_descriptor_page(m_space->id, xdes_n);
+      if (descr_page_no != last_descr_page_no)
+        last_descr= buf_page_get_gen(page_id_t{m_space->id, xdes_n},
+                                     0, RW_S_LATCH, nullptr,
+                                     BUF_GET_POSSIBLY_FREED, &mtr,
+                                     &err);
+      xdes_t *descr= XDES_ARR_OFFSET + XDES_SIZE
+        * xdes_calc_descriptor_index(0, xdes_n)
+	  + last_descr->page.frame;
+      last_descr_page_no= descr_page_no;
+      if (xdes_n % srv_page_size == 0 && xdes_get_n_used(descr) == 2)
+        continue;
+      m_extent_info[xdes_n]= uint32_t(xdes_get_state(descr));
+    }
+    mtr.commit();
+  }
+
+  /** Find the earlier free extent for the given used extent
+  @param max_limit Find the extent below max limit extent
+  @return value or FIL_NULL if there is no extent */
+  uint32_t find_free_extent(uint32_t max_limit)
+  {
+    for (auto &extent_info : m_extent_info)
+    {
+      if (max_limit <= extent_info.first)
+        return FIL_NULL;
+
+      /* Ignore the already used extents */
+      if (extent_info.second >> 31)
+        continue;
+      if (extent_info.second == XDES_FREE)
+      {
+        /* Mark the extent as used one */
+        extent_info.second |= (1U << 31);
+        return extent_info.first;
+      }
+    }
+    return FIL_NULL;
+  }
+
+  /** Find the index from the given table for the given index id
+  @param table table to be searched
+  @param index_id index id to be search
+  @return index */
+  dict_index_t *find_index(const dict_table_t *table, uint64_t index_id)
+  {
+    for (dict_index_t *idx= dict_table_get_first_index(table);
+         idx; idx= dict_table_get_next_index(idx))
+      if (idx->id == index_id)
+        return idx;
+    return nullptr;
+  }
+
+  /** Find the index on the system tables for the given index_id */
+  dict_index_t* find_sys_index(uint64_t index_id)
+  {
+    dict_index_t *idx= find_index(dict_sys.sys_tables, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_columns, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_indexes, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_fields, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_foreign, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_foreign_cols, index_id);
+    if (!idx)
+      idx= find_index(dict_sys.sys_virtual, index_id);
+    return idx;
+  }
+
+  /** Defragment the indexes */
+  dberr_t defragment_index(dict_index_t *index)
+  {
+    ib::info() <<"Defragmentation of index " << index->name()
+               << " in table " << index->table->name << " happens";
```

Comment:
The function is missing `noexcept`.

For output, we should prefer `sql_print_information`. The grammar could be simplified:
```c++
sql_print_information("Defragmenting index %`s of table %`.*s.%`s", …);
```